### PR TITLE
COMMERCE-9006 As a developer I want to unify all renderers for Product Details page - 2/2 (Frontend Story)

### PR DESCRIPTION
**Pull request by lorenzolapescara**
https://issues.liferay.com/browse/COMMERCE-9006

### CODE REVIEW

**Comment by andrea-ale-sbarra** [comment]
@lorenzolapescara I think you should move the last 3 commits of @christian-chiappa to be the last 3 commits of this PR.

Thanks!

File: modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceProducts.macro [hunk 1]
```diff
@@ -1,5 +1,23 @@
 definition {
 
+	macro addEntriesToGroupedProduct {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceProducts.macro:3]: Rename it to addGroupedProductEntries

File: modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceProducts.macro [hunk 2]
```diff
@@ -1,5 +1,23 @@
...
+		Click(locator1 = "Button#PLUS");
+
+		SelectFrame.selectFrameNoLoading(locator1 = "IFrame#MODAL_BODY");
+
+		for (var productName : list "${productsList}") {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceProducts.macro:8]: productsList -> productsNames

File: modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceProducts.macro [hunk 3]
```diff
@@ -585,6 +603,76 @@ definition {
 			value1 = "\RETURN");
 	}
 
+	macro setProductInventoryConfiguration {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceProducts.macro:600]: updateProductInventoryConfiguration

File: modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceProducts.macro [hunk 4]
```diff
@@ -585,6 +603,76 @@ definition {
...
+			Type(
+				key_labelName = "Minimum Order Quantity",
+				locator1 = "TextInput#SPECIFIC_LABEL_TEXT_INPUT",
+				value1 = "${minOrderQuantity}");
+		}
+
+		Button.clickPublish();
+	}
+
+	macro setProductVirtualConfiguration {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceProducts.macro:650]: addProductVirtualConfiguration

File: modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceProducts.macro [hunk 5]
```diff
@@ -593,4 +681,33 @@ definition {
 		Button.clickPublish();
 	}
 
+	macro setVirtualProductSample {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceProducts.macro:678]: uploadVirtualProductSample

File: modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceUtilities.macro [hunk 6]
```diff
@@ -6,4 +6,54 @@ definition {
...
 
+	macro convertStringToKey {
+		var string = StringUtil.lowerCase("${string}");
+
+		var string = StringUtil.replace("${string}", " ", "-");
+
+		return "${string}";
+	}
+
+	macro getFutureDateIncreaseYears {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceUtilities.macro:17]: increaseDateYears

**Comment by lorenzolapescara** [comment]
ci:test:commerce-functional

File: modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceUtilities.macro [hunk 7]
```diff
@@ -6,4 +6,54 @@ definition {
...
+		else {
+			var futureDateYear = "${currentDateYear}";
+		}
+
+		var futureDate = "${currentDateMonth}/${currentDateDay}/${futureDateYear}";
+
+		return "${futureDate}";
+	}
+
+	macro selectAndUploadSingleFile {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceUtilities.macro:40]: uploadSingleFile

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase [hunk 8]
```diff
@@ -478,6 +479,203 @@ definition {
 		}
 	}
 
+	@description = "COMMERCE-9674 | As a buyer, I want to be able to view the incoming dates/quantities for a multiple SKUs product"
+	@priority = "3"
+	test CanViewIncomingQuantityForMultipleSkusProduct {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase:641]: CanViewMultipleSkusProductIncomingQuantity

File: modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceUtilities.macro [hunk 9]
```diff
@@ -6,4 +6,54 @@ definition {
...
+		}
+
+		if (isSet(increaseYears)) {
+			var futureDateYear = ${currentDateYear} + ${increaseYears};
+		}
+		else {
+			var futureDateYear = "${currentDateYear}";
+		}
+
+		var futureDate = "${currentDateMonth}/${currentDateDay}/${futureDateYear}";
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceUtilities.macro:35]: Get current value only if a value is not set

File: modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceUtilities.macro [hunk 10]
```diff
@@ -6,4 +6,54 @@ definition {
...
+		else {
+			var currentDateYear = DateUtil.getCurrentYear();
+		}
+
+		if (isSet(increaseYears)) {
+			var futureDateYear = ${currentDateYear} + ${increaseYears};
+		}
+		else {
+			var futureDateYear = "${currentDateYear}";
+		}
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceUtilities.macro:33]:
```
if (isSet(increaseYears)) {
			var dateYear = ${dateYear} + ${increaseYears};
		}

Comment (alessiorendina) [reply modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceUtilities.macro:33]: Same for months and days

File: modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceUtilities.macro [hunk 11]
```diff
@@ -6,4 +6,54 @@ definition {
...
+	macro getFutureDateIncreaseYears {
+		var currentDateDay = DateUtil.getCurrentDate();
+		var currentDateMonth = DateUtil.getCurrentMonth();
+
+		if (isSet(yearFormat)) {
+			var currentDateYear = DateUtil.getFormattedCurrentDate("${yearFormat}");
+		}
+		else {
+			var currentDateYear = DateUtil.getCurrentYear();
+		}
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/macros/CommerceUtilities.macro:26]: Remove "current" prefix

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase [hunk 12]
```diff
@@ -478,6 +479,203 @@ definition {
...
+			AssertElementNotPresent(
+				key_incomingDate = "${incomingFormattedDate2}",
+				key_incomingQuantity = "2",
+				locator1 = "CommerceFrontStore#FS_PRODUCT_DETAIL_INCOMING_DATE_QUANTITY");
+		}
+	}
+
+	@description = "COMMERCE-9673 | As a buyer, I want to be able to view the incoming date/quantity for a single SKU product"
+	@priority = "3"
+	test CanViewIncomingQuantityForSingleSkuProduct {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase:781]: CanViewSingleSkuProductIncomingQuantity

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase [hunk 13]
```diff
@@ -478,6 +479,203 @@ definition {
...
+				urlAppend = "web/minium",
+				userEmailAddress = "buyer@liferay.com");
+
+			CommerceAcceleratorsInitializer.changeProductsSorting(sortingType = "Name Ascending");
+
+			CommerceAcceleratorsInitializer.gotoMiniumProductDetailsPage(productName = "ABS Sensor");
+		}
+
+		task ("Then the buyer can see the incoming date/quantity with the nearest date") {
+			var nearestIncomingFormattedDate =  CommerceUtilities.getFutureDateIncreaseYears(increaseYears = "2", yearFormat = "yy");
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase]: nearestIncomingDate

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase [hunk 14]
```diff
@@ -478,6 +479,203 @@ definition {
...
+				locator1 = "CommerceFrontStore#FS_PRODUCT_SKU");
+
+			AssertElementPresent(
+				key_incomingDate = "${incomingFormattedDate1}",
+				key_incomingQuantity = "4",
+				locator1 = "CommerceFrontStore#FS_PRODUCT_DETAIL_INCOMING_DATE_QUANTITY");
+		}
+
+		task ("And the buyer can select the second SKU and see the incoming quantity") {
+			var incomingFormattedDate2 = CommerceUtilities.getFutureDateIncreaseYears(increaseYears = "3", yearFormat = "yy");
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase]: incomingDate2

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase [hunk 15]
```diff
@@ -478,6 +479,203 @@ definition {
...
+				locator1 = "AppBuilder#SEARCH_BAR_INPUT",
+				value1 = "Hoses");
+
+			Click(locator1 = "Button#SUBMIT");
+
+			CommerceAcceleratorsInitializer.gotoMiniumProductDetailsPage(productName = "Hoses");
+		}
+
+		task ("Then the buyer can select the first SKU and see the incoming quantity with the nearest date") {
+			var incomingFormattedDate1 = CommerceUtilities.getFutureDateIncreaseYears(increaseYears = "2", yearFormat = "yy");
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase]: incomingDate1

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase [hunk 16]
```diff
@@ -483,6 +483,78 @@ definition {
 		}
 	}
 
+	@description = "COMMERCE-9675. Users can view and download a product's attachments"
+	@priority = "3"
+	test CanViewAndDownloadProductAttachments {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase:569]: CanDownloadProductAttachments

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase [hunk 17]
```diff
@@ -680,6 +752,78 @@ definition {
 		}
 	}
 
+	@description = "COMMERCE-9675. Users can view and download a product's attachments"
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase]: This is a duplicate, please remove it

File: modules/apps/commerce/commerce-product-test/src/testFunctional/macros/FrontStore.macro [hunk 18]
```diff
@@ -319,6 +414,32 @@ definition {
 		return "${quantityConfigurationText}";
 	}
 
+	macro viewGroupedProductEntries {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/macros/FrontStore.macro:417]: Check if this is a duplicate

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase [hunk 19]
```diff
@@ -588,9 +588,12 @@ definition {
 
 			Navigator.gotoNavTab(navTab = "Incoming");
 
-			var nearestIncomingDate = CommerceUtilities.getFutureDateIncreaseYears(increaseYears = "2", yearFormat = "yyyy");
-			var fartherIncomingDate = CommerceUtilities.getFutureDateIncreaseYears(increaseYears = "3", yearFormat = "yyyy");
-
+			var nearestIncomingDate = CommerceUtilities.getFutureDateIncreaseYears(
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase:672]: Check if this is a duplicate or just SF. Please move it under the same commit.

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase [hunk 20]
```diff
@@ -779,16 +791,90 @@ definition {
 				userSiteMembership = "Minium");
 		}
 
-		task ("And a product with 2 attachments") {
+		task ("And a new Grouped product is created with all the data needed") {
+			for (var productName : list "${groupedProductEntries}") {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase:876]: `for (var groupedProductEntryName : list "${groupedProductEntryNames}") {`

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase [hunk 21]
```diff
@@ -779,16 +791,90 @@ definition {
...
+				salePrice = "50.00");
 
 			CommerceEntry.gotoMenuTab(menuTab = "Media");
 
-			CommerceProductImages.addAttachments(fileNames = "${fileName1},${fileName2}");
+			CommerceProductImages.addProductImages(imageTitles = "Black,Blue");
+
+			CommerceEntry.gotoMenuTab(menuTab = "Subscription");
+
+			for (var subscriptionName : list "Payment,Delivery") {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase:930]: subscriptionName -> subscriptionType

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase [hunk 22]
```diff
@@ -483,6 +483,87 @@ definition {
 		}
 	}
 
+	@description = "COMMERCE-9678 | As a buyer, I want to not be able to see Full Description tab if the field is empty"
+	@priority = "3"
+	test CannotViewFullDescriptionTabWhenFieldIsEmpty {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CommerceProductDetails.testcase:488]: CanAssertFullDescriptionTabNotPresent

**Review by alessiorendina — CHANGES_REQUESTED** [review]
@lorenzolapescara please check my comments, thanks!

File: modules/apps/commerce/commerce-product-type-grouped-web/src/main/java/com/liferay/commerce/product/type/grouped/web/internal/util/GroupedCPTypeHelperImpl.java [hunk 23]
```diff
@@ -36,15 +39,27 @@ public class GroupedCPTypeHelperImpl implements GroupedCPTypeHelper {
 
 	@Override
 	public List<CPDefinitionGroupedEntry> getCPDefinitionGroupedEntry(
+			long commerceAccountId, long commerceChannelGroupId,
 			long cpDefinitionId)
 		throws PortalException {
 
-		return _cpDefinitionGroupedEntryService.getCPDefinitionGroupedEntries(
+
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-type-grouped-web/src/main/java/com/liferay/commerce/product/type/grouped/web/internal/util/GroupedCPTypeHelperImpl.java:46]: Remove empty line

File: modules/apps/commerce/commerce-product-type-grouped-web/src/main/java/com/liferay/commerce/product/type/grouped/web/internal/util/GroupedCPTypeHelperImpl.java [hunk 24]
```diff
@@ -36,15 +39,27 @@ public class GroupedCPTypeHelperImpl implements GroupedCPTypeHelper {
...
-		return _cpDefinitionGroupedEntryService.getCPDefinitionGroupedEntries(
+		
+		if(_commerceProductViewPermission.contains(
+			PermissionThreadLocal.getPermissionChecker(), commerceAccountId,
+			commerceChannelGroupId, cpDefinitionId)) {
+		return _cpDefinitionGroupedEntryLocalService.getCPDefinitionGroupedEntries(
 			cpDefinitionId, QueryUtil.ALL_POS, QueryUtil.ALL_POS,
 			new CPDefinitionGroupedEntryPriorityComparator());
+		}
+		return Collections.emptyList();
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-type-grouped-web/src/main/java/com/liferay/commerce/product/type/grouped/web/internal/util/GroupedCPTypeHelperImpl.java:54]: There should be an empty line before this one.

File: modules/apps/commerce/commerce-product-type-grouped-web/src/main/java/com/liferay/commerce/product/type/grouped/web/internal/util/GroupedCPTypeHelperImpl.java [hunk 25]
```diff
@@ -36,15 +39,27 @@ public class GroupedCPTypeHelperImpl implements GroupedCPTypeHelper {
...
+			long commerceAccountId, long commerceChannelGroupId,
 			long cpDefinitionId)
 		throws PortalException {
 
-		return _cpDefinitionGroupedEntryService.getCPDefinitionGroupedEntries(
+		
+		if(_commerceProductViewPermission.contains(
+			PermissionThreadLocal.getPermissionChecker(), commerceAccountId,
+			commerceChannelGroupId, cpDefinitionId)) {
+		return _cpDefinitionGroupedEntryLocalService.getCPDefinitionGroupedEntries(
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-product-type-grouped-web/src/main/java/com/liferay/commerce/product/type/grouped/web/internal/util/GroupedCPTypeHelperImpl.java:50]: There should be an empty line before this one.

**Review by alessiorendina — CHANGES_REQUESTED** [review]
@lorenzolapescara I also left some SF comments, please make sure to run SF locally, thanks.

File: modules/apps/commerce/commerce-product-test/src/testFunctional/macros/FrontStore.macro (hunk 18 above)
Comment (lorenzolapescara) [reply modules/apps/commerce/commerce-product-test/src/testFunctional/macros/FrontStore.macro:417]: This is not a duplicate: 2 new macros have been added that are related to the grouped products: `CommerceProducts.addEntriesToGroupedProduct()` and `FrontStore.viewGroupedProductEntries()`.
Anyway, I'm going to move macro to the "Add macros" commit

**Comment by lorenzolapescara** [comment]
ci:test:sf

**Comment by lorenzolapescara** [comment]
ci:test:sf

**Comment by lorenzolapescara** [comment]
ci:test:commerce

**Comment by lorenzolapescara** [comment]
ci:test:sf

**Comment by lorenzolapescara** [comment]
ci:test:relevant

**Comment by lorenzolapescara** [comment]
QA and PM review passed. Forwarding

**Comment by lorenzolapescara** [comment]
ci:forward

**Pull request closed without merging**